Expand on Snapshot and the function parameter (#63)
tc39/proposal-async-context: README.md

  // AsyncContext.Variable was restored after the previous run.
  console.log(asyncVar.get()); // => 'top'
}

function randomTimeout() {
  return Math.random() * 1000;
}
```

> Note: There are controversial thought on the dynamic scoping and
> `Variable`, checkout [SCOPING.md][] for more details.

Hosts are expected to use the infrastructure in this proposal to allow tracking
not only asynchronous callstacks, but other ways to schedule jobs on the event
loop (such as `setTimeout`) to maximize the value of these use cases.

A detailed example of use cases can be found in the
[Use cases document](./USE-CASES.md).

## `AsyncContext.Snapshot`

`Snapshot` allows you to opaquely capture the current values of all `Variable`s

> and execute a function at a later time with as if those values were still the

Reviewer: ```suggestion
and execute a function at a later time as if those values were still the
```

current values (a snapshot and restore).

Note that even with `Snapshot`, you can only access the value associated with

> an `Variable` instance if you have access to that instance.

Reviewer: ```suggestion
a `Variable` instance if you have access to that instance.
```

the values of all `AsyncContext.Variable`s without access to any of them. This
allows the user-land queue to be implemented in a way that is decoupled from
consumers of `AsyncContext.Variable`.

```typescript
// The scheduler doesn't access to any AsyncContext.Variable.
const scheduler = {
  queue: [],
  postTask(task) {
    // Each callback is stored with the context at which it was enqueued.
    const snapshot = new AsyncContext.Snapshot();
    queue.push(() => snapshot.run(task));
  },
  runWhenIdle() {
    // All callbacks in the queue would be run with the current context if they
    // hadn't been wrapped.
    for (const cb of this.queue) {
      cb();
    }
    this.queue = [];
  }
};

function userAction() {
  scheduler.postTask(function userTask() {
    console.log(traceContext.get());
  });
}

// Tracing libraries can use AsyncContext.Variable to store tracing contexts.
const traceContext = new AsyncContext.Variable();
traceContext.run("trace-id-a", userAction);
traceContext.run("trace-id-b", userAction);

runWhenIdle();
// The userTask will be run with the trace context it was enqueued with.
// => 'trace-id-a'
// => 'trace-id-b'
```

# FAQ

> ## Why takes a function in `run`?

Reviewer: ```suggestion
## Why take a function in `run`?
```

The `Variable.prototype.run` and `Snapshot.prototype.run` methods take a

> function to execute because it is the a way to ensure async context variables

Reviewer: ```suggestion
function to execute because it ensures async context variables
```

> will always be consistent values in a given execution flow. Any modification

Reviewer: ```suggestion
will always contain consistent values in a given execution flow. Any modification
```

must be taken in a sub-graph of an async execution flow, and can not affect

> their parent scopes.

Reviewer: ```suggestion
their parent or sibling scopes.
```
